text/0000-macro-metavar-expr.md

```diff
@@ -0,0 +1,332 @@
+preallocate the vector with the correct length.  However, there is no standard
+facility in declarative macros to achieve this, as there is no way to obtain
+the *number* of repetitions of `$x`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The [example `vec` macro defininition in the guide][guide-vec] could be made
+more efficient if it could use `Vec::with_capacity` to pre-allocate a vector
+with the correct capacity.  To do this, we need to know the number of
+repetitions.
+
+[guide-vec]: https://doc.rust-lang.org/book/ch19-06-macros.html#declarative-macros-with-macro_rules-for-general-metaprogramming
+
+Metadata about metavariables, like the number of repetitions, can be accessed
+using **metavariable expressions**.  The metavariable expression for the
+count of the number of repetitions of a metavariable `x` is `${count(x)}`, so
+we can improve the `vec` macro as follows:
+
+```
+#[macro_export]
+macro_rules! vec {
+    ( $( $x:expr ),* ) => {
+        {
+            let mut temp_vec = Vec::with_capacity(${count(x)});
+            $(
+                temp_vec.push($x);
+            )*
+            temp_vec
+        }
+    };
+}
+```
+
+The following metavariable expressions are available:
+
+| Expression                 | Meaning    |
+|----------------------------|------------|
+| `${count(ident)}`          | The number of times `$ident` repeats in total. |
+| `${count(ident, depth)}`   | The number of times `$ident` repeats at up to `depth` nested repetition depths. |
+| `${index()}`               | The current index of the inner-most repetition. |
+| `${index(depth)}`          | The current index of the nested repetition at `depth` steps out. |
+| `${length()}`              | The length of the inner-most repetition. |
+| `${length(depth)}`         | The length of the nested repetition at `depth` steps out. |
+| `${ignore(ident)}`         | Binds `$ident` for repetition, but expands to nothing. |
+| `$$`                       | Expands to a single `$`, for removing ambiguity in recursive macro definitions. |
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Metavariable expressions in declarative macros provide expansions for
+information about metavariables that are otherwise not easily obtainable.
+
+This is a backwards-compatible change as both `$$` and `${ .. }` are not
+currently accepted as valid.
+
+The metavariable expressions added in this RFC are concerned with declarative
+macro metavariable repetitions, and obtaining the information that the
+compiler knows about the repetitions that are being processed.
+
+## Count
+
+The `${count(x)}` metavariable expression shown in the `vec` example in the
+previous section counts the number of repetitions that will occur if the
+identifier is used in a repetition at this depth.  This means that in a macro
+expansion like:
+
+```
+    ${count(x)} $( $x )*
```

Comment:
Just to clarify: is `${count(x)}` (and more generally, `${count(x,d)}` legal to write *inside* a `$(  ... $x ... )*` repetition block?

Assuming it is legal there, does it expand the same way regardless of whether it is outside or inside such a `$( ... $x ... )*` block?

Reply:
I guess metavariable expressions in general *have* to be legal *inside* a `$( ... $x ... )*` repetition block, since that's the only way that `${index()}` metavariable expressions make sense.

Reply:
Also, the Larger example section does show a demo of `${count(x)}` and `${count(x,d)}` that clearly shows the values varying. So that's great.

And the text does say "number of repetitions that will occur [...] **at this depth**". So my concern was just me not reading carefully enough the whole way through.